src/vit-based/utilities.jl

```diff
@@ -0,0 +1,19 @@
+using LinearAlgebra: mul!
+
+chunk(A, k::Int; dim::Int = 1) = 
+    (selectdim(A, dim, i) for i in Iterators.partition(axes(A,dim), cld(size(A,dim), k)));
+
+pair(t) = (isa(t, Pair)) ? t : (t, t)
+
+function batchmul!(C, A, B)
+    for j in axes(A, 4), i in axes(A, 3)
+        @views mul!(C[:, :, i, j], A[:, :, i, j], B[:, :, i, j])
```

Comment:
This does not seem AD friendly

Reply:
NNlib has some batched ops.

Reply:
Good call, I think `batched_mul` should be updated to support >3D arrays.

Reply:
As a stop-gap, perhaps tensorcast?

Reply:
Is `batched_mul` even needed if MHA is implemented using `Parallel`? I suppose there's a performance argument, but the current implementation isn't doing that.

Reply:
TensorCast appears to be doing a naive dot-product, I think:
```julia
julia> @pretty @reduce z[h, w, c, n] := sum(i) x[h, i, c, n] * y[i, w, c, n]
begin
    @boundscheck ndims(x) == 4 || throw(ArgumentError("expected a 4-tensor x[h, i, c, n]"))
    @boundscheck axes(x, 2) == axes(y, 1) || throw(DimensionMismatch("range of index i must agree"))
    @boundscheck axes(x, 3) == axes(y, 3) || throw(DimensionMismatch("range of index c must agree"))
    @boundscheck axes(x, 4) == axes(y, 4) || throw(DimensionMismatch("range of index n must agree"))
    @boundscheck ndims(y) == 4 || throw(ArgumentError("expected a 4-tensor y[i, w, c, n]"))
    local nightingale = transmute(x, Val((1, nothing, 3, 4, 2)))
    local waterbuffalo = transmute(y, Val((nothing, 2, 3, 4, 1)))
    z = dropdims(sum(@__dot__(nightingale * waterbuffalo), dims = 5), dims = 5)
end
```
@mcabbott can correct me if I wrote this operation improperly for TensorCast.

Reply:
Yes that's correct. All `@reduce` knows how to do is to broadcast out & sum. Not always as terrible as it sounds but probably not the fastest.

The reason `batched_mul` doesn't yet support 4d arrays is being uncertain about how best to handle batch dims. At present it broadcasts e.g. `3,3,3 ⊠ 3,3,1 -> 3,3,3`, but if it wants to handle e.g. `3,3,3,3 ⊠ 3,3,1,3 -> 3,3,3,3` then it's not just a reshape to the 3d case.

Reply:
Something like https://github.com/chengchingwen/Transformers.jl/blob/master/src/fix/batchedmul.jl might help, but it seems to be for 3D tensors still

Reply:
https://github.com/chengchingwen/Transformers.jl/blob/master/src/fix/batchedmul.jl#L13 appears to accept higher-dimensional inputs. I can't tell whether it's GPU friendly, however.

Reply:
IIRC the code in Transformers might be an ancestor of that in NNlib? Both call CUDA routines, e.g. https://github.com/chengchingwen/Transformers.jl/blob/master/src/cuda/batch_gemm_gpu.jl, which I think only handle 3d arrays.

Reply:
Exploring further led me to https://github.com/chengchingwen/NeuralAttentionlib.jl, which seems to have quite some stuff implemented - but it might mean taking on an additional dep

Reply:
Unfortunately still seems to be 3D arrays🥲

Reply:
Does https://github.com/chengchingwen/NeuralAttentionlib.jl/blob/master/src/matmul/matmul.jl fit the bill? I'm not familiar enough with the internals to comment 😅

Reply:
@mcabbott based on the commit dates I think so. It wouldn't be too bad to just copy https://github.com/chengchingwen/Transformers.jl/blob/master/src/fix/batchedmul.jl#L13-L23 over for this PR and make it use `NNlib.batched_mul` instead. Keeping it as an internal function sidesteps the ambiguity question.

@theabhirath worth a try I'd say, though it's not clear to me whether those functions handle >3D inputs. If they do, we could consider adding it as a dep to get an MHA layer for free as well. JuliaHub tells me that only one additional transitive dep (PartialFunctions, which is relatively lightweight) would be added.

Reply:
I stand corrected, the MHA layer itself is relegated to [test code](https://github.com/chengchingwen/NeuralAttentionlib.jl/blob/f389adb71714f4245233e2e8e97292ebbac6148b/test/old_impl/mh_atten.jl#L35). However, the workhorse functions that power its implementation should be available as part of NeuralAttentionLib.

Reply:
Yeah they're still for 3D inputs though, it's not going to help here 🥲 I'll try and see if I can find something else